### PR TITLE
Type visualization response contract

### PR DESCRIPTION
**Pull request by mohavro**
## Architectural Alignment

- Backend: FastAPI production path response typing is changed.
- Frontend: no changes.
- Gradio: no changes.

Mapped current shapes before editing:

- `get_3d_visualization_data_enhanced()` returns `(positions, asset_ids, colors, hover)`; sample: `position=[1.0, 0.0, 0.0]`, `asset_id='AAPL'`, `color='#4ECDC4'`, `hover='Asset: AAPL'`.
- `GET /api/visualization` returns `nodes` with `id`, `symbol`, `name`, `asset_class`, `x`, `y`, `z`, `color`, `size`; `edges` with `source`, `target`, `relationship_type`, `strength`.

Field classification:

- Domain: `id`, `symbol`, `asset_class`, `source`, `target`, `relationship_type`, `strength`.
- Presentation: `name`, `x`, `y`, `z`, `color`, `size`.

## Primary Objective

Make the visualization API response a typed contract so domain fields and presentation encodings are explicit and extra fields are rejected.

### In Scope

- Add `VisualizationNode` and `VisualizationEdge` Pydantic response models matching the current API payload.
- Forbid extra node, edge, and root response fields via `ConfigDict(extra="forbid")`.
- Type `size` as `int` to match the actual payload.
- Change `VisualizationDataResponse` from `list[dict[str, Any]]` to typed node/edge lists.
- Update existing response-model tests to use the real current payload shape.
- Add an API shape enforcement test for exact node/edge key sets.
- Add a graph-helper test proving visualization generation does not mutate metrics.

### Out of Scope

- Graph algorithms.
- Layout logic.
- Density logic.
- Pagination.
- Frontend changes.
- Field renaming.

### Files Expected to Change

- `api/api_models.py` — defines typed visualization response models.
- `tests/unit/test_api_main.py` — enforces API visualization response shape.
- `tests/unit/test_asset_graph.py` — verifies visualization helper does not mutate graph metrics.

## Validation Commands

```bash
DATABASE_URL="sqlite:///:memory:" SECRET_KEY="test-secret" ADMIN_USERNAME="admin" ADMIN_PASSWORD="password" PATH="/workspace/.venv/bin:$PATH" python - <<'PY'
# shape inspection script used before edits
PY
PATH="/workspace/.venv/bin:$PATH" pytest tests/unit/test_api_main.py -v
PATH="/workspace/.venv/bin:$PATH" pytest tests/unit/test_asset_graph.py -v
PATH="/workspace/.venv/bin:$PATH" pytest tests/unit/test_api.py -v
git diff --check -- "api/api_models.py" "tests/unit/test_api_main.py" "tests/unit/test_asset_graph.py"
PATH="/workspace/.venv/bin:$PATH" pre-commit run --files "api/api_models.py" "tests/unit/test_api_main.py"
```

Shape inspection, pytest, and `git diff --check` passed. `pre-commit` ran; formatting/import hooks passed, while existing flake8-docstrings in `tests/unit/test_api_main.py` and unrelated mypy errors in `src/data/sample_data.py` remain outside this scoped change.

## Merge Criteria

- [x] Scope is tightly aligned to the Primary Objective
- [x] Required pytest validation passes locally
- [x] Changes align with production architecture (FastAPI + Next.js)

### Scope Compliance

- [x] This PR makes one primary decision only (see Primary Objective)
- [x] I have explicitly listed what is out of scope
- [x] Runtime behavior changes are limited to typed visualization response validation
- [x] I have verified the branch, base branch, and referenced PR/commit/ref context before concluding merge status or PR necessity
- [x] I have checked this PR against the production architecture (`FastAPI` backend + `Next.js` frontend)
- [x] I have checked this PR against `.github/AUTOMATION_SCOPE_POLICY.md`

**Related Documentation**:

- [PR Scope Guardrails](../docs/PR_SCOPE_GUARDRAILS.md)
- [Automation Scope Policy](./AUTOMATION_SCOPE_POLICY.md)

## Additional Notes

Codacy MCP tools were not available in this workspace, so Codacy analysis could not be run for the edited files.

<div><a href="https://cursor.com/agents/bc-91b717a0-9c33-4cd8-a348-3aae36ae1678"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-web-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-web-light.png">Open in Web</picture></a>&nbsp;<a href="https://cursor.com/background-agent?bcId=bc-91b717a0-9c33-4cd8-a348-3aae36ae1678"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-cursor-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-cursor-light.png">Open in Cursor</picture></a>&nbsp;</div>

## Summary by cubic
Typed the visualization API response in `FastAPI` with `pydantic` models for nodes and edges, forbidding extra fields (including at the root) and enforcing exact field sets. Tests ensure non-empty nodes/edges, valid edge strength, and stable metrics; payload shape is unchanged and no frontend changes.

- **Refactors**
  - Added `VisualizationNode`/`VisualizationEdge` with `ConfigDict(extra="forbid")`; `VisualizationDataResponse` now uses them, forbids extra root fields, and `size` is `int`.
  - Expanded tests: enforce exact node/edge keys, require non-empty lists in `/api/visualization`, check edge strength range, and verify the visualization helper doesn’t mutate metrics.

<sup>Written for commit 491e0658ece3cb3f1992f9cc6617a5d6504117de. Summary will update on new commits. <a href="https://cubic.dev/pr/DashFin-FarDb/financial-asset-relationship-db/pull/1079?utm_source=github">Review in cubic</a></sup>

___

## **CodeAnt-AI Description**
**Make visualization API responses follow a strict typed shape**

### What Changed
- Visualization responses now require the full set of node and edge fields used by the graph view, including asset details and visual values.
- Extra fields are rejected in visualization nodes, edges, and the top-level response, so the API returns only the expected payload.
- Tests now verify the live visualization response has non-empty nodes and edges with exact field sets.
- Added coverage to ensure generating visualization data does not change graph metrics.

### Impact
`✅ Clearer visualization API responses`
`✅ Fewer broken graph payloads`
`✅ Safer graph rendering for consumers`

[🔄 Retrigger CodeAnt AI Review](https://api.codeant.ai/pr/retrigger_review?id=SPv1CyV85xbYHnxOowINV7Slm3wylOhkhfbd-kj4XsM&org=DashFin-FarDb)<details>
<summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

### CODE REVIEW

**Comment by mohavro** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scan Engine |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/dashfin.fardb/pr-checks/a9cf479f-3a05-4caa-a649-233f72959275?refs=491e0658ece3cb3f1992f9cc6617a5d6504117de&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

**Review by mohavro — COMMENTED** [review]
Phase 3 visualization schema review:

The PR is correctly scoped and implements the intended visualization contract seam.

Confirmed:

- `VisualizationNode` and `VisualizationEdge` now define the typed API contract instead of `list[dict[str, Any]]`.
- `VisualizationDataResponse` now uses typed lists:
  - `nodes: list[VisualizationNode]`
  - `edges: list[VisualizationEdge]`
- The model fields match the current `/api/visualization` payload described in the PR:
  - node fields: `id`, `symbol`, `name`, `asset_class`, `x`, `y`, `z`, `color`, `size`
  - edge fields: `source`, `target`, `relationship_type`, `strength`
- Tests now enforce domain/presentation field presence in the API response.
- The graph-helper test verifies visualization generation does not mutate `calculate_metrics()` output.
- No density, metrics ownership, pagination, frontend, CORS/auth/settings, or layout-algorithm changes are included.

One non-blocking observation:

`VisualizationNode.size` is typed as `float`, while the response-model test uses `size: 5`. Pydantic will coerce this to float, so this is not a blocker. If the runtime payload actually emits integer sizes, `float` is still acceptable because numeric JSON consumers can handle it and the field represents a presentation scalar. Do not widen this unless a consumer/test requires it.

Merge recommendation:

Safe to mark ready and merge once required CI checks are green. After merge, update #1028 and close/update #1078 as the visualization node/edge schema contract complete. The remaining Phase 3 seam is the asset pagination contract decision.

**Comment by mohavro** [comment]
@codeant-ai review (manual request)

**Comment by mohavro** [comment]
@coderabbitai review PR

**Comment by mohavro** [comment]
@gemini-code-assist review PR

**Comment by mohavro** [comment]
@codex review PR

**Review by mohavro — COMMENTED** [review]
Follow-up review on the new comments:

The comments are valid and worth addressing before merge because they strengthen the actual contract enforcement without widening scope.

Required / recommended changes:

## 1. Strengthen the shape-enforcement test

The new test currently checks a subset of fields. Since the PR objective is a typed visualization response contract, make the test assert the complete current node/edge key sets.

Use this pattern:

```python
node_keys = {"id", "symbol", "name", "asset_class", "x", "y", "z", "color", "size"}
edge_keys = {"source", "target", "relationship_type", "strength"}

assert all(node_keys.issubset(node) for node in viz_data["nodes"])
assert all(edge_keys.issubset(edge) for edge in viz_data["edges"])
```

This keeps the test concise and makes it actually validate both domain and presentation fields.

## 2. Forbid extra fields on `VisualizationNode` and `VisualizationEdge`

This is the most important contract-enforcement improvement.

Right now the models are typed, but Pydantic defaults can still allow/ignore unexpected extra fields depending on config/inheritance. For an explicit response contract, add `extra="forbid"` to both nested models.

Expected shape:

```python
from pydantic import BaseModel, ConfigDict

class VisualizationNode(BaseModel):
    """Response model for a visualization node."""

    model_config = ConfigDict(extra="forbid")

    ...

class VisualizationEdge(BaseModel):
    """Response model for a visualization edge."""

    model_config = ConfigDict(extra="forbid")

    ...
```

This makes accidental payload drift fail validation rather than being silently accepted.

## 3. Align `VisualizationNode.size` with the actual payload

If the visualization router currently emits integer sizes via `min`/`max` over integer constants, type the response model accurately:

```python
size: int
```

Do not introduce `PositiveInt` or bounds in this PR unless existing code/tests already enforce those bounds. `int` is enough to align the OpenAPI contract with the current payload.

## 4. Keep scope tight

Do not change field names, layout logic, graph algorithms, frontend code, or pagination here.

## Validation to run after these changes

```bash
python -m pytest tests/unit/test_api_main.py -v
python -m pytest tests/unit/test_asset_graph.py -v
python -m pytest tests/unit/test_api.py -v
pre-commit run --files api/api_models.py tests/unit/test_api_main.py tests/unit/test_asset_graph.py
```

Once those three changes are made and CI is green, this PR should be safe to merge.

**Review by mohavro — COMMENTED** [review]
Additional test-design fix before merge:

The strengthened key-set assertions are good, but they should be guarded by explicit non-empty checks. Otherwise `all(...)` is vacuously true if the visualization endpoint returns empty `nodes` or `edges`, which would make the contract test pass without validating any actual item shape.

Please update `test_visualization_response_enforces_domain_and_presentation_shape` so it asserts the fixture produced at least one node and one edge before checking the key sets.

Recommended final shape:

```python
def test_visualization_response_enforces_domain_and_presentation_shape(self, client: TestClient) -> None:
    """Visualization payload should expose typed domain fields and presentation encodings."""
    response = client.get("/api/visualization")
    assert response.status_code == 200
    viz_data = response.json()

    assert "nodes" in viz_data
    assert "edges" in viz_data
    assert len(viz_data["nodes"]) > 0, "Graph must have nodes for shape verification"
    assert len(viz_data["edges"]) > 0, "Graph must have edges for shape verification"

    node_keys = {"id", "symbol", "name", "asset_class", "x", "y", "z", "color", "size"}
    edge_keys = {"source", "target", "relationship_type", "strength"}

    assert all(node_keys.issubset(node) for node in viz_data["nodes"])
    assert all(edge_keys.issubset(edge) for edge in viz_data["edges"])
```

This keeps the test scoped to the current seeded graph behaviour and avoids a false-positive contract test on empty lists.

No runtime code change is implied by this comment.

File: api/api_models.py [hunk 1]
```diff
@@ -40,8 +40,31 @@ class MetricsResponse(BaseModel):
     relationship_density: float = 0.0
 
 
+class VisualizationNode(BaseModel):
```
[the remark this answers is not in the document]
Comment (mohavro) [reply api/api_models.py:43]: Valid follow-up, but out of scope for this shape-contract PR. Current PR defines fields and forbids extras; presentation value-format validation can be handled separately if needed.

**Review by mohavro — COMMENTED** [review]
merge PR approved
